### ISSUE TITLE
Implement polynomial simplifier with replacements

### ISSUE THREAD

**Issue by WhiteBlackGoose** [asc-community/AngouriMath#185]
e^x + sin(x) + 2e^x + 2sin(x) =>
a + b + 2a + 2b
=> P(a) = 3a + CONST where CONST = b + 2b = Q(b) = 3b
=> 3a + 3b
=> 3e^x + 3sin(x)

**Comment by WhiteBlackGoose** [comment]
- [ ] Entity's parent && GetMinimumSubtree using parents
- [ ] Step-by-step replacer

**Comment by Happypig375** [comment]
Isn't this what's already nearly implemented?

**Comment by Rafael-SOWNet** [comment]
Yes. Measured on `master` at `2f89ab36`, the example from the body and its neighbours:

```csharp
"e^x + sin(x) + 2e^x + 2sin(x)".ToEntity().Simplify()          // 3 * (sin(x) + e ^ x)
"e^x + sin(x) + 2 e^x + 2 sin(x) - 3 e^x".ToEntity().Simplify() // 3 * sin(x)
"a + b + 2 a + 2 b".ToEntity().Simplify()                       // 3 * (a + b)
"x^2 y + 3 x^2 y - x^2 y".ToEntity().Simplify()                 // 3 * x ^ 2 * y
```

Like terms are collected whatever the term is — a function, a power, a product — without going through a named replacement, because the collector keys on the term itself. `3 * (sin(x) + e^x)` is `3 e^x + 3 sin(x)` with the common factor taken out, which `Simplify` prefers as the smaller tree. The 2020 checklist's third item, "a little bit faster than infinity per operation", is `SimplifyHard` at 107 ms on today's desktop.

What the replacement idea still buys is *factoring* a polynomial in a subexpression — `sin(x)^2 + 2 sin(x) + 1` is left as written by `Factorize` today while `x^2 + 2 x + 1` gives `(x + 1)^2` — and that is the one open case of #203, which I am doing there. Closing this one as delivered.
